ext/calendar/jewish.c

```diff
@@ -433,16 +433,25 @@ static void MoladOfMetonicCycle(
 								   zend_long *pMoladHalakim)
 {
 	register zend_ulong r1, r2, d1, d2;
+	zend_long chk;
 
 	/* Start with the time of the first molad after creation. */
 	r1 = NEW_MOON_OF_CREATION;
+	chk = (zend_long)metonicCycle;
+
+	if (chk > (ZEND_LONG_MAX / (HALAKIM_PER_METONIC_CYCLE & 0xFFFF)) ||
+	    chk > (ZEND_LONG_MAX / ((HALAKIM_PER_METONIC_CYCLE >> 16) & 0xFFFF))) {
+		*pMoladDay = 0;
+		*pMoladHalakim = 0;
+		return;
+	}
 
 	/* Calculate metonicCycle * HALAKIM_PER_METONIC_CYCLE.  The upper 32
 	 * bits of the result will be in r2 and the lower 16 bits will be
 	 * in r1. */
-	r1 += metonicCycle * (HALAKIM_PER_METONIC_CYCLE & 0xFFFF);
+	r1 += chk * (HALAKIM_PER_METONIC_CYCLE & 0xFFFF);
```

Comment:
This can still overflow, because r1 starts with NEW_MOON_OF_CREATION.
Analogous for r2.